File: src/main/java/com/polykhel/billnext/service/query/ActivityQueryService.java
package com.polykhel.billnext.service.query;

import com.polykhel.billnext.domain.*;
import com.polykhel.billnext.repository.ActivityRepository;
import com.polykhel.billnext.service.criteria.ActivityCriteria;
import com.polykhel.billnext.service.dto.ActivityDTO;
import com.polykhel.billnext.service.mapper.ActivityMapper;
import java.util.List;
import javax.persistence.criteria.JoinType;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.data.jpa.domain.Specification;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import tech.jhipster.service.QueryService;

/**
 * Service for executing complex queries for {@link Activity} entities in the database.
 * The main input is a {@link ActivityCriteria} which gets converted to {@link Specification},
 * in a way that all the filters must apply.
 * It returns a {@link List} of {@link ActivityDTO} or a {@link Page} of {@link ActivityDTO} which fulfills the criteria.
 */
@Service
@Transactional(readOnly = true)
public class ActivityQueryService extends QueryService<Activity> {

    private final Logger log = LoggerFactory.getLogger(ActivityQueryService.class);

    private final ActivityRepository activityRepository;

    private final ActivityMapper activityMapper;

    public ActivityQueryService(ActivityRepository activityRepository, ActivityMapper activityMapper) {
        this.activityRepository = activityRepository;
        this.activityMapper = activityMapper;
    }

    /**
     * Return a {@link List} of {@link ActivityDTO} which matches the criteria from the database.
     * @param criteria The object which holds all the filters, which the entities should match.
     * @return the matching entities.
     */
    @Transactional(readOnly = true)
    public List<ActivityDTO> findByCriteria(ActivityCriteria criteria) {
        log.debug("find by criteria : {}", criteria);
        final Specification<Activity> specification = createSpecification(criteria);
        return activityMapper.toDto(activityRepository.findAll(specification));
    }

    /**
     * Return a {@link Page} of {@link ActivityDTO} which matches the criteria from the database.
     * @param criteria The object which holds all the filters, which the entities should match.
     * @param page The page, which should be returned.
     * @return the matching entities.
     */
    @Transactional(readOnly = true)
    public Page<ActivityDTO> findByCriteria(ActivityCriteria criteria, Pageable page) {
        log.debug("find by criteria : {}, page: {}", criteria, page);
        final Specification<Activity> specification = createSpecification(criteria);
        return activityRepository.findAll(specification, page).map(activityMapper::toDto);
    }

    /**
     * Return the number of matching entities in the database.
     * @param criteria The object which holds all the filters, which the entities should match.
     * @return the number of matching entities.
     */
    @Transactional(readOnly = true)
    public long countByCriteria(ActivityCriteria criteria) {
        log.debug("count by criteria : {}", criteria);
        final Specification<Activity> specification = createSpecification(criteria);
        return activityRepository.count(specification);
    }

    /**
     * Function to convert {@link ActivityCriteria} to a {@link Specification}
     * @param criteria The object which holds all the filters, which the entities should match.
     * @return the matching {@link Specification} of the entity.
     */
    protected Specification<Activity> createSpecification(ActivityCriteria criteria) {
        Specification<Activity> specification = Specification.where(null);
        if (criteria != null) {
            if (criteria.getId() != null) {
                specification = specification.and(buildRangeSpecification(criteria.getId(), Activity_.id));
            }
            if (criteria.getDate() != null) {
                specification = specification.and(buildRangeSpecification(criteria.getDate(), Activity_.date));
            }
            if (criteria.getAmount() != null) {
                specification = specification.and(buildRangeSpecification(criteria.getAmount(), Activity_.amount));
            }
            if (criteria.getRemarks() != null) {
                specification = specification.and(buildStringSpecification(criteria.getRemarks(), Activity_.remarks));
            }
            if (criteria.getType() != null) {
                specification = specification.and(buildSpecification(criteria.getType(), Activity_.type));
            }
            if (criteria.getUserId() != null) {
                specification =
                    specification.and(
                        buildSpecification(criteria.getUserId(), root -> root.join(Activity_.user, JoinType.LEFT).get(User_.id))
                    );
            }
            if (criteria.getWalletId() != null) {
                specification =
                    specification.and(
                        buildSpecification(criteria.getWalletId(), root -> root.join(Activity_.wallet, JoinType.LEFT).get(Wallet_.id))
                    );
            }
            if (criteria.getCategoryId() != null) {
                specification =
                    specification.and(
                        buildSpecification(criteria.getCategoryId(), root -> root.join(Activity_.category, JoinType.LEFT).get(Category_.id))
                    );
            }
        }
        return specification;
    }
}

File: src/main/kotlin/com/polykhel/billnext/service/ActivityQueryService.kt
package com.polykhel.billnext.service

import com.polykhel.billnext.domain.*
import com.polykhel.billnext.repository.ActivityRepository
import com.polykhel.billnext.service.criteria.ActivityCriteria
import com.polykhel.billnext.service.dto.ActivityDTO
import com.polykhel.billnext.service.mapper.ActivityMapper
import org.slf4j.LoggerFactory
import org.springframework.data.domain.Page
import org.springframework.data.domain.Pageable
import org.springframework.data.jpa.domain.Specification
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import tech.jhipster.service.QueryService
import tech.jhipster.service.filter.Filter
import javax.persistence.criteria.JoinType

/**
 * Service for executing complex queries for [Activity] entities in the database.
 * The main input is a [ActivityCriteria] which gets converted to [Specification],
 * in a way that all the filters must apply.
 * It returns a [MutableList] of [ActivityDTO] or a [Page] of [ActivityDTO] which fulfills the criteria.
 */
@Service
@Transactional(readOnly = true)
class ActivityQueryService(
    private val activityRepository: ActivityRepository,
    private val activityMapper: ActivityMapper
) : QueryService<Activity>() {

    private val log = LoggerFactory.getLogger(javaClass)

    /**
     * Return a [MutableList] of [ActivityDTO] which matches the criteria from the database.
     * @param criteria The object which holds all the filters, which the entities should match.
     * @return the matching entities.
     */
    @Transactional(readOnly = true)
    fun findByCriteria(criteria: ActivityCriteria?): MutableList<ActivityDTO> {
        log.debug("find by criteria : $criteria")
        val specification = createSpecification(criteria)
        return activityMapper.toDto(activityRepository.findAll(specification))
    }

    /**
     * Return a [Page] of [ActivityDTO] which matches the criteria from the database.
     * @param criteria The object which holds all the filters, which the entities should match.
     * @param page The page, which should be returned.
     * @return the matching entities.
     */
    @Transactional(readOnly = true)
    fun findByCriteria(criteria: ActivityCriteria?, page: Pageable): Page<ActivityDTO> {
        log.debug("find by criteria : $criteria, page: $page")
        val specification = createSpecification(criteria)
        return activityRepository.findAll(specification, page)
            .map(activityMapper::toDto)
    }

    /**
     * Return the number of matching entities in the database.
     * @param criteria The object which holds all the filters, which the entities should match.
     * @return the number of matching entities.
     */
    @Transactional(readOnly = true)
    fun countByCriteria(criteria: ActivityCriteria?): Long {
        log.debug("count by criteria : $criteria")
        val specification = createSpecification(criteria)
        return activityRepository.count(specification)
    }

    /**
     * Function to convert [ActivityCriteria] to a [Specification].
     * @param criteria The object which holds all the filters, which the entities should match.
     * @return the matching [Specification] of the entity.
     */
    protected fun createSpecification(criteria: ActivityCriteria?): Specification<Activity?> {
        var specification: Specification<Activity?> = Specification.where(null)
        if (criteria != null) {
            if (criteria.id != null) {
                specification = specification.and(buildRangeSpecification(criteria.id, Activity_.id))
            }
            if (criteria.date != null) {
                specification = specification.and(buildRangeSpecification(criteria.date, Activity_.date))
            }
            if (criteria.amount != null) {
                specification = specification.and(buildRangeSpecification(criteria.amount, Activity_.amount))
            }
            if (criteria.remarks != null) {
                specification = specification.and(buildStringSpecification(criteria.remarks, Activity_.remarks))
            }
            if (criteria.type != null) {
                specification = specification.and(buildSpecification(criteria.type, Activity_.type))
            }
            if (criteria.userId != null) {
                specification = specification.and(
                    buildSpecification(criteria.userId as Filter<Long>) {
                        it.join(Activity_.user, JoinType.LEFT).get(User_.id)
                    }
                )
            }
            if (criteria.walletId != null) {
                specification = specification.and(
                    buildSpecification(criteria.walletId as Filter<Long>) {
                        it.join(Activity_.wallet, JoinType.LEFT).get(Wallet_.id)
                    }
                )
            }
            if (criteria.categoryId != null) {
                specification = specification.and(
                    buildSpecification(criteria.categoryId as Filter<Long>) {
                        it.join(Activity_.category, JoinType.LEFT).get(Category_.id)
                    }
                )
            }
        }
        return specification
    }
}
